File: projects/pyrift/targets/pattern_engine.py
# pyrift/targets/pattern_engine.py
# Polar Bipartite Pattern Matching Engine - Python Implementation

import re
import threading
from typing import List, Optional, Tuple, Dict, Callable
from dataclasses import dataclass, field
from enum import IntEnum


class PatternPolarity(IntEnum):
    LEFT = 0   # Input/matcher
    RIGHT = 1  # Output/generator


@dataclass
class RiftPattern:
    """Compiled pattern with metadata"""
    pattern_str: str
    polarity: PatternPolarity
    priority: int
    anchored: bool
    is_literal: bool
    compiled_regex: Optional[re.Pattern] = None


@dataclass
class BipartitePair:
    """Links left pattern (input) to right pattern (output)"""
    left: RiftPattern
    right: RiftPattern
    transform_fn: Optional[Callable] = None
    is_governed: bool = False
    transform_id: int = 0


@dataclass
class MatchResult:
    """Result of pattern matching"""
    matched: bool
    output: Optional[str] = None
    priority: int = 0
    transform_id: int = 0
    groups: Dict[str, str] = field(default_factory=dict)


class PatternEngine:
    """
    Polar Bipartite Pattern Matching Engine
    Left patterns match input using regex
    Right patterns generate output from templates
    """
    
    def __init__(self, mode: str = "classical"):
        self.pairs: List[BipartitePair] = []
        self.mode = mode
        self._lock = threading.RLock()
        
        # Metrics
        self.total_matches = 0
        self.total_failures = 0
        self.average_match_time_ms = 0.0
    
    def add_pair(self, left_pattern: str, right_pattern: str, 
                 priority: int = 100, right_is_literal: bool = False) -> bool:
        """Add a bipartite pattern pair"""
        with self._lock:
            # Create left pattern (input matcher)
            left = RiftPattern(
                pattern_str=left_pattern,
                polarity=PatternPolarity.LEFT,
                priority=priority,
                anchored=left_pattern.startswith('^'),
                is_literal=False
            )
            
            # Compile left regex
            try:
                left.compiled_regex = re.compile(left_pattern)
            except re.error:
                return False
            
            # Create right pattern (output generator)
            right = RiftPattern(
                pattern_str=right_pattern,
                polarity=PatternPolarity.RIGHT,
                priority=priority,
                anchored=False,
                is_literal=right_is_literal
            )
            
            # Compile right if it's a regex
            if not right_is_literal:
                try:
                    right.compiled_regex = re.compile(right_pattern)
                except re.error:
                    right.is_literal = True
            
            # Create pair
            pair = BipartitePair(
                left=left,
                right=right,
                transform_id=len(self.pairs) + 1
            )
            
            self.pairs.append(pair)
            return True
    
    def match(self, input_str: str) -> MatchResult:
        """Match input against all left patterns, return best match"""
        import time
        
        start_time = time.time()
        
        with self._lock:
            best_pair: Optional[BipartitePair] = None
            best_priority = float('inf')
            best_match = None
            
            # Search for matching pattern (respecting priority)
            for pair in self.pairs:
                if pair.left.compiled_regex:
                    match = pair.left.compiled_regex.search(input_str)
                    if match:
                        # Lower priority number = higher priority
                        if pair.left.priority < best_priority:
                            best_pair = pair
                            best_priority = pair.left.priority
                            best_match = match
            
            # Generate output
            if best_pair:
                template = best_pair.right.pattern_str
                
                if best_pair.right.is_literal:
                    output = template
                else:
                    # Simple substitution for capture groups
                    output = template
                    if best_match:
                        for i, group in enumerate(best_match.groups() or [], 1):
                            output = output.replace(f'${i}', group or '')
                        # Named groups
                        for name, value in (best_match.groupdict() or {}).items():
                            output = output.replace(f'{{{name}}}', value or '')
                
                # Update metrics
                elapsed = (time.time() - start_time) * 1000
                self.total_matches += 1
                self._update_metrics(elapsed)
                
                return MatchResult(
                    matched=True,
                    output=output,
                    priority=best_priority,
                    transform_id=best_pair.transform_id,
                    groups=best_match.groupdict() if best_match else {}
                )
            
            # No match found
            elapsed = (time.time() - start_time) * 1000
            self.total_failures += 1
            self._update_metrics(elapsed)
            
            return MatchResult(matched=False)
    
    def _update_metrics(self, elapsed_ms: float) -> None:
        """Update running average match time"""
        total = self.total_matches + self.total_failures
        if total > 0:
            self.average_match_time_ms = (
                (self.average_match_time_ms * (total - 1)) + elapsed_ms
            ) / total
    
    def get_metrics(self) -> Dict[str, any]:
        """Get engine metrics"""
        with self._lock:
            return {
                'total_matches': self.total_matches,
                'total_failures': self.total_failures,
                'average_match_time_ms': self.average_match_time_ms,
                'pair_count': len(self.pairs)
            }


# Example usage patterns for Python code generation
DEFAULT_PYTHON_PATTERNS = [
    # Function definition transformation
    (r'def\s+(\w+)\s*\(\s*\)\s*:', r'def \1(self):', 100),
    # Class inheritance
    (r'class\s+(\w+)\s*:', r'class \1(RiftObject):', 90),
    # Variable annotation
    (r'(\w+)\s*:=\s*(.+)', r'self.\1 = \2', 80),
    # Quantum decorator
    (r'@quantum', r'@rift_quantum_decorator', 200),
]


def create_default_engine() -> PatternEngine:
    """Create pattern engine with default Python transformations"""
    engine = PatternEngine()
    for left, right, priority in DEFAULT_PYTHON_PATTERNS:
        engine.add_pair(left, right, priority, right_is_literal=True)
    return engine

File: bindings/lua-riftlang/pattern_engine.lua
-- luarift/targets/pattern_engine.lua
-- Polar Bipartite Pattern Matching Engine - Lua Implementation

local pattern_engine = {}

-- Pattern polarity
pattern_engine.POLARITY = {
    LEFT = 0,   -- Input/matcher
    RIGHT = 1   -- Output/generator
}

-- ============================================================================
-- RiftPattern Class
-- ============================================================================

local RiftPattern = {}
RiftPattern.__index = RiftPattern

function RiftPattern.new(pattern_str, polarity, priority, anchored, is_literal)
    local self = setmetatable({}, RiftPattern)
    self.pattern_str = pattern_str
    self.polarity = polarity
    self.priority = priority
    self.anchored = anchored
    self.is_literal = is_literal
    self.compiled_pattern = nil
    return self
end

pattern_engine.RiftPattern = RiftPattern

-- ============================================================================
-- BipartitePair Class
-- ============================================================================

local BipartitePair = {}
BipartitePair.__index = BipartitePair

function BipartitePair.new(left, right, transform_fn, is_governed, transform_id)
    local self = setmetatable({}, BipartitePair)
    self.left = left
    self.right = right
    self.transform_fn = transform_fn
    self.is_governed = is_governed or false
    self.transform_id = transform_id or 0
    return self
end

pattern_engine.BipartitePair = BipartitePair

-- ============================================================================
-- MatchResult Class
-- ============================================================================

local MatchResult = {}
MatchResult.__index = MatchResult

function MatchResult.new(matched, output, priority, transform_id, groups)
    local self = setmetatable({}, MatchResult)
    self.matched = matched
    self.output = output
    self.priority = priority or 0
    self.transform_id = transform_id or 0
    self.groups = groups or {}
    return self
end

pattern_engine.MatchResult = MatchResult

-- ============================================================================
-- PatternEngine Class
-- ============================================================================

local PatternEngine = {}
PatternEngine.__index = PatternEngine

function PatternEngine.new(mode)
    local self = setmetatable({}, PatternEngine)
    self.pairs = {}
    self.mode = mode or "classical"
    
    -- Metrics
    self.total_matches = 0
    self.total_failures = 0
    self.average_match_time_ms = 0.0
    
    return self
end

function PatternEngine:add_pair(left_pattern, right_pattern, priority, right_is_literal)
    priority = priority or 100
    right_is_literal = right_is_literal or false
    
    -- Create left pattern (input matcher)
    local left = RiftPattern.new(
        left_pattern,
        pattern_engine.POLARITY.LEFT,
        priority,
        left_pattern:sub(1, 1) == "^",
        false
    )
    
    -- Create right pattern (output generator)
    local right = RiftPattern.new(
        right_pattern,
        pattern_engine.POLARITY.RIGHT,
        priority,
        false,
        right_is_literal
    )
    
    -- Create pair
    local pair = BipartitePair.new(
        left,
        right,
        nil,
        false,
        #self.pairs + 1
    )
    
    table.insert(self.pairs, pair)
    return true
end

function PatternEngine:match(input_str)
    local start_time = os.clock()
    
    local best_pair = nil
    local best_priority = math.huge
    local best_match = nil
    local best_groups = {}
    
    -- Search for matching pattern (respecting priority)
    for _, pair in ipairs(self.pairs) do
        -- Try to match input against left pattern
        local captures = {input_str:match(pair.left.pattern_str)}
        if #captures > 0 then
            -- Lower priority number = higher priority
            if pair.left.priority < best_priority then
                best_pair = pair
                best_priority = pair.left.priority
                best_match = captures
            end
        end
    end
    
    -- Generate output
    if best_pair then
        local template = best_pair.right.pattern_str
        local output = template
        
        -- Simple substitution for capture groups
        if best_match then
            for i, capture in ipairs(best_match) do
                output = output:gsub("$" .. i, capture)
            end
        end
        
        -- Update metrics
        local elapsed = (os.clock() - start_time) * 1000
        self.total_matches = self.total_matches + 1
        self:_update_metrics(elapsed)
        
        return MatchResult.new(
            true,
            output,
            best_priority,
            best_pair.transform_id,
            best_groups
        )
    end
    
    -- No match found
    local elapsed = (os.clock() - start_time) * 1000
    self.total_failures = self.total_failures + 1
    self:_update_metrics(elapsed)
    
    return MatchResult.new(false)
end

function PatternEngine:_update_metrics(elapsed_ms)
    local total = self.total_matches + self.total_failures
    if total > 0 then
        self.average_match_time_ms = (
            (self.average_match_time_ms * (total - 1)) + elapsed_ms
        ) / total
    end
end

function PatternEngine:get_metrics()
    return {
        total_matches = self.total_matches,
        total_failures = self.total_failures,
        average_match_time_ms = self.average_match_time_ms,
        pair_count = #self.pairs
    }
end

pattern_engine.PatternEngine = PatternEngine

-- ============================================================================
-- Default Lua Patterns
-- ============================================================================

pattern_engine.DEFAULT_LUA_PATTERNS = {
    -- Function definition transformation
    {"function%s+(%w+)%s*%(%)%s*", "function %1(self) end", 100},
    -- Local variable annotation
    {"local%s+(%w+)%s*=%s*(.+)", "rift_local('%1', %2)", 90},
    -- Table creation
    {"(%w+)%s*=%s*{%s*}", "rift_table('%1', {})", 80},
    -- Quantum decorator
    {"@quantum", "@rift_quantum_decorator", 200},
}

function pattern_engine.create_default_engine()
    local engine = PatternEngine.new()
    for _, pattern in ipairs(pattern_engine.DEFAULT_LUA_PATTERNS) do
        engine:add_pair(pattern[1], pattern[2], pattern[3], true)
    end
    return engine
end

return pattern_engine
